File: backend/app/utils/url_normalize.py
from __future__ import annotations

import re
from urllib.parse import urlsplit, urlunsplit

_URL_IN_TEXT_RE = re.compile(r"https?://[^\s<>\"'\[\](){}]+", re.IGNORECASE)
_TRAILING_PUNCT_RE = re.compile(r"[,，。.;；!?！？)、】』」\]\/]+$", re.UNICODE)


class UrlNormalizeError(ValueError):
    pass


def _strip_trailing_punctuation(value: str) -> str:
    s = value.strip()
    while s and _TRAILING_PUNCT_RE.search(s):
        s = _TRAILING_PUNCT_RE.sub("", s)
    return s


def _extract_first_url(text: str) -> str | None:
    match = _URL_IN_TEXT_RE.search(text)
    if not match:
        return None
    return _strip_trailing_punctuation(match.group(0))


def _is_bilibili_host(hostname: str) -> bool:
    host = (hostname or "").lower().rstrip(".")
    return host.endswith("bilibili.com") or host == "b23.tv" or host.endswith(".b23.tv")


def _is_douyin_host(hostname: str) -> bool:
    host = (hostname or "").lower().rstrip(".")
    return host.endswith("douyin.com")


def _is_x_host(hostname: str) -> bool:
    host = (hostname or "").lower().rstrip(".")
    return host.endswith("twitter.com") or host.endswith("x.com")


def _is_mp4_url(url: str, path: str, query: str) -> bool:
    combined = f"{path}{query}"
    return bool(re.search(r"\.mp4(\?|#|$)", combined, re.IGNORECASE)) or bool(
        re.search(r"\.mp4", url, re.IGNORECASE)
    )


def _is_x_status_path(path: str) -> bool:
    return bool(re.search(r"/status/\d+", path or "", re.IGNORECASE))


def _normalize_bilibili(parts) -> str:
    path = (parts.path or "").rstrip("/") or ""
    if re.match(r"^/video/BV[\w-]+$", path, re.IGNORECASE):
        path = f"{path}/"
    elif path.startswith("/video/"):
        path = path
    new_parts = parts._replace(query="", fragment="", path=path)
    return urlunsplit(new_parts)


def _normalize_douyin(parts) -> str:
    path = parts.path or "/"
    if not path.endswith("/"):
        path = f"{path}/"
    new_parts = parts._replace(query="", fragment="", path=path)
    return urlunsplit(new_parts)


def _normalize_x(parts) -> str:
    new_parts = parts._replace(query="", fragment="")
    return urlunsplit(new_parts)


def normalize_video_url(raw: str) -> str:
    """Normalize user-pasted share text or URLs for video extraction."""
    trimmed = (raw or "").strip()
    if not trimmed:
        raise UrlNormalizeError("empty url")

    candidate = _extract_first_url(trimmed)
    if not candidate and re.match(r"^https?://", trimmed, re.IGNORECASE):
        candidate = _strip_trailing_punctuation(trimmed)
    if not candidate:
        raise UrlNormalizeError("no url found in input")

    parts = urlsplit(candidate)
    if not parts.scheme or not parts.netloc:
        raise UrlNormalizeError("invalid url")

    host = parts.hostname or ""
    url_for_check = candidate

    if _is_mp4_url(url_for_check, parts.path or "", parts.query or ""):
        return candidate
    if _is_bilibili_host(host):
        return _normalize_bilibili(parts)
    if _is_douyin_host(host):
        return _normalize_douyin(parts)
    if _is_x_host(host) and _is_x_status_path(parts.path or ""):
        return _normalize_x(parts)
    return candidate

File: frontend/src/utils/urlNormalize.js
const URL_IN_TEXT_RE = /https?:\/\/[^\s<>"'[\](){}]+/i
const TRAILING_PUNCT_RE = /[,，。.;；!?！？)、】』」\]\/]+$/u

function stripTrailingPunctuation(value) {
  let s = String(value || '').trim()
  while (s && TRAILING_PUNCT_RE.test(s)) {
    s = s.replace(TRAILING_PUNCT_RE, '')
  }
  return s
}

function extractFirstUrl(text) {
  const match = String(text || '').match(URL_IN_TEXT_RE)
  if (!match) {
    return null
  }
  return stripTrailingPunctuation(match[0])
}

function parseUrlSafe(url) {
  try {
    return new URL(url)
  } catch {
    return null
  }
}

function isBilibiliHost(hostname) {
  const host = String(hostname || '').toLowerCase()
  return /(^|\.)bilibili\.com$/i.test(host) || /(^|\.)b23\.tv$/i.test(host)
}

function isDouyinHost(hostname) {
  const host = String(hostname || '').toLowerCase()
  return /(^|\.)douyin\.com$/i.test(host)
}

function isXHost(hostname) {
  const host = String(hostname || '').toLowerCase()
  return /(^|\.)twitter\.com$/i.test(host) || /(^|\.)x\.com$/i.test(host)
}

function isMp4Url(urlString, pathname, search) {
  const combined = `${pathname || ''}${search || ''}`
  return /\.mp4(\?|#|$)/i.test(combined) || /\.mp4/i.test(urlString)
}

function isXStatusPath(pathname) {
  return /\/status\/\d+/i.test(pathname || '')
}

function normalizeBilibiliUrl(parsed) {
  const path = (parsed.pathname || '').replace(/\/+$/, '') || ''
  parsed.search = ''
  parsed.hash = ''
  if (/^\/video\/BV[\w-]+$/i.test(path)) {
    parsed.pathname = `${path}/`
  }
  return parsed.toString()
}

function normalizeDouyinModalId(parsed) {
  const modalId = parsed.searchParams.get('modal_id')
  if (modalId && /^\d+$/.test(modalId)) {
    return `https://www.iesdouyin.com/share/video/${modalId}/`
  }
  return null
}

function normalizeDouyinUrl(parsed) {
  const modalNormalized = normalizeDouyinModalId(parsed)
  if (modalNormalized) {
    return modalNormalized
  }

  const path = parsed.pathname || '/'
  if (/^\/video\/\d+$/.test(path)) {
    parsed.search = ''
    parsed.hash = ''
    return parsed.toString()
  }

  parsed.search = ''
  parsed.hash = ''
  parsed.pathname = path.endsWith('/') ? path : `${path}/`
  return parsed.toString()
}

function normalizeXUrl(parsed) {
  parsed.search = ''
  parsed.hash = ''
  return parsed.toString()
}

/**
 * @param {string} raw
 * @returns {{ url: string, changed: boolean, error?: string }}
 */
export function normalizeUserVideoInput(raw) {
  const trimmed = String(raw ?? '').trim()
  if (!trimmed) {
    return { url: '', changed: false, error: 'empty' }
  }

  let candidate = extractFirstUrl(trimmed)
  if (!candidate && /^https?:\/\//i.test(trimmed)) {
    candidate = stripTrailingPunctuation(trimmed)
  }
  if (!candidate) {
    return { url: '', changed: false, error: 'no_url' }
  }

  const parsed = parseUrlSafe(candidate)
  if (!parsed) {
    return { url: '', changed: false, error: 'invalid_url' }
  }

  const host = parsed.hostname
  let normalized = candidate

  if (isMp4Url(candidate, parsed.pathname, parsed.search)) {
    normalized = candidate
  } else if (isBilibiliHost(host)) {
    normalized = normalizeBilibiliUrl(parsed)
  } else if (isDouyinHost(host)) {
    normalized = normalizeDouyinUrl(parsed)
  } else if (isXHost(host) && isXStatusPath(parsed.pathname)) {
    normalized = normalizeXUrl(parsed)
  } else {
    normalized = candidate
  }

  return {
    url: normalized,
    changed: normalized !== trimmed,
    error: undefined,
  }
}
